### DISCUSSION TITLE
Swap operator

### DISCUSSION THREAD

**Discussion by deeprobin** [dotnet/csharplang#5622] [Language Ideas]
A swap operator would be very nice. This should of course only be used when 2 values are to be swapped. 
Local variables can probably be optimized away by the JIT anyway, but if you access references this operator becomes very interesting imo.

# Current way to swap variables:
```cs
var a = 1;
var b = 2;

var c = a;
a = b;
b = c;
```

# Proposed operator
```cs
var a = 3;
var b = 4;

a <-> b;
```

## Alternative designs
```cs
a <=> b;
```

**Comment by CyrusNajmabadi** [comment]
`(b, a) = (a, b);`

This is already idiomatic c# (and the ide will detect the long form and suggest this). 

> `a <=> b;`

I would like to reserve the spaceship operator for 3-way comparisons :-)

**Reply by ichensky** [reply]
I would prefer to have `swap` function for this 
```csharp
swap(arr[i][node[j]], values[i]);
```
looks better and easier for understanding than 
```csharp
(values[i], arr[i][node[j]]) = (arr[i][node[j]], values[i]);
```

**Reply by CyrusNajmabadi** [reply]
You can write your own helper that does exactly that.  :)   no need for the language to have it.  If you want the BCL to have it, the request should go to dotnet/runtime.

**Reply by HaloFour** [reply]
```cs
public static void swap<T>(ref T x, ref T y) {
    (x, y) = (y, x);
}
```

And you can call it like:

```cs
swap(ref arr[i][node[j]], ref values[i]);
```

**Reply by deeprobin** [reply]
@CyrusNajmabadi
We could also use the swap operator only to explicitly swap the contents in the memory and allow this operator only in the unsafe context.

I suggested the spaceship operator only as an alternative. In my opinion `<->` is important in the semantic context.
However, this is (like pretty much any language feature) just something subjective.
For example, I find an swap like `(a, b) = (b, a)` less readable than `a <-> b`.

@ichensky @HaloFour
I honestly find a swap function for this unnecessary. The `(a, b) = (b, a)` variant is imo faster to write.

**Reply by CyrusNajmabadi** [reply]
> We could also use the swap operator only to explicitly swap the contents in the memory and allow this operator only in the unsafe context.

This is so niche as to be totally unnecessary.  You can just write your own helpers for this.  

> For example, I find an swap like (a, b) = (b, a) less readable than a <-> b.

At a certain point it's not really wise or desirable to just keep try to take more and more niche operations and to give them terser and terser syntaxes.

**Reply by tannergooding** [reply]
Worth noting there is a minor codegen inefficiency around `(b, a) = (a, b)` compared to `var tmp = a; a = b; b = tmp;`

Would be nice to get that fixed so they are equivalent.

-- And noting that in some cases this isn't something the JIT can do due to potential side-effects, it's something Roslyn likely needs to optimize in its IL output

**Reply by CyrusNajmabadi** [reply]
@jcouv as i believe he implemented tuple swap.

**Reply by PathogenDavid** [reply]
Roslyn already emits `(b, a) = (a, b)` as using temporaries rather than actual tuples. [It uses an extra local](https://sharplab.io/#v2:EYLgtghglgdgPgAQEwEYCwAoBBmABM3AYUwG9NcL89YAXXCAbnMp11t2CY0t2YtYQAWXAFkAFAEo+PXAF4AfLjHAANPQlylENcAlce0gcJFJJ0stxkUAbhABOuGmAAOmxtJ4RNnD5WCanZ31KAF9MEKA), but the JIT output is the same (in this situation at least.)

**Reply by tannergooding** [reply]
Its not the same for `refs` and other scenarios.

See https://github.com/dotnet/runtime/pull/64412#discussion_r857798132 where we explicitly backed out using it because the real world codegen was worse.

**Reply by PathogenDavid** [reply]
Oops, my bad. I wasn't paying enough attention to the context of this subthread. That's good to know though.

**Comment by Rekkonnect** [comment]
I've been trying to wrap my head around the existence of such a feature, and to me it appears the best option is to use a `swap` keyword. The syntax would be as simple as `swap a, b`, which would be a full statement, and be the equivalent of calling a BCL method, defined as such:
```csharp
public static class Swapping
{
    public static void Swap<T>(ref T a, ref T b)
    {
        (a, b) = (b, a);
    }
}
```

The feature would be expandable by transforming the `swap` statement into a `swap` expression, and changing the return type of the method into the resulting tuple. I personally cannot find much use in getting the resulting tuple from the expression directly, so I would prefer to stick into the simple `swap` statement, if it is ever to be considered.

**Reply by zms9110750** [reply]
Please take care of non-native English speakers. We prefer symbols to words.

**Reply by Rekkonnect** [reply]
With 100 keywords already, I'm confident enough that's not barely any concern. You can't tell me that `swap` as a word is not common, or simple, compared to others that exist like `yield`, `interface`, etc.

**Reply by zms9110750** [reply]
Man, because there are already 100 keywords, that's the problem.
In fact, I really don't know what yield interface means.I rely on intellicode to write code. Many times, I only need to remember the first three letters.
Get, set, value, protected only appear under certain circumstances. But it seems that swap can appear anywhere. If I often pop up something I don't understand but can't use when writing, I will be really troubled.
I even want to use ":" instead of "in" in foreach

**Reply by CyrusNajmabadi** [reply]
@zms9110750 all these things need to be learned, regardless of if they're keywords or punctuation. 

> Man, because there are already 100 keywords, that's the problem.

How would it make it better if that were 100 choices of punctuation instead?

> In fact, I really don't know what yield interface means.

If you don't know what `interface` means, then that will make using and writing c# very hard. It's a very core concept. 

> I rely on intellicode to write code. 

This will make no difference if something is a keyword or operator. Intellicode suggests both. 

> Get, set, value, protected only appear under certain circumstances. But it seems that swap can appear anywhere. 

All these would appear where legal. If something is legal in more places it will appear in more places. 

> IIf I often pop up something I don't understand but can't use when writing, I will be really troubled.

Best thing for that would be to spend some time learning, and then you won't have the issue if not understanding something offered. 

> I even want to use ":" instead of "in" in foreach

This is a good example. This would make no substantive difference. So if a keyword bothers you here then that's something you likely need to find a way to address on your end. C# blends keywords and punctuation. It's deeply part of the style of our language.
